### PR TITLE
[release][website] Add release notes for Release 2.4.2

### PR DESCRIPTION
**Pull request by wolfstudy**
Signed-off-by: xiaolong.ran <rxl@apache.org>

### Modifications

- Add release notes for Release 2.4.2.

### CODE REVIEW

**Comment by wolfstudy** [comment]
@sijie @jiazhai @codelipenghui @merlimat PTAL, thanks

File: site2/website/release-notes.md [hunk 1]
```diff
@@ -1,6 +1,109 @@
 
 ## Apache
 
+### 2.4.2 &mdash; 2019-12-04 <a id="2.4.2"></a>
+
+#### Fixes
+
+* Fixed don't set interrupt flag again after catching interrupt exception in Pulsar Client #5643
```
Comment (codelipenghui) [inline site2/website/release-notes.md]: Please ensure #5643 can link to the PR

File: site2/website/release-notes.md [hunk 2]
```diff
@@ -1,6 +1,109 @@
...
+* Fixed message corruption on OOM for batch messages #5443
+* Fixed couple functions related integration tests #5434
+* Fixed bug that namespace policies does not take effect due to NPE #5408
+* Fixed race condition : Failed to read-more entries on dispatcher #5391
+* Fixed potential deadlock that can occur in addConsumer #5371
+* Fixed proxy to be able to re-send request body #5361
+* Fixed pulsar can't load the customized SerDe #5357
+* Fixed instability in Pulsar Function window integration test #5337
+* Fixed bk write failure part 2 #5322
+* Allow to specify delivery delay in C++ client #5317
```
Comment (codelipenghui) [inline site2/website/release-notes.md]: It looks like a enhancement

File: site2/website/release-notes.md [hunk 3]
```diff
@@ -1,6 +1,109 @@
...
+* Fixed windowed functions were broken when we changed java function instance to use classloaders #5117
+* Fixed storage size always 0 without subscription #5108
+
+#### Enhancements
+
+* Add subscribe position param for consumer of sink #5532
+* Efficiency improvements for delay delivery tracker #5498
+* Add is_read_compacted to create_reader() in python API #5483
+* Make some common use method of ManagedLedger public #5472
+* Print unknown exception stacktrace #5444
```
Comment (codelipenghui) [inline site2/website/release-notes.md]: I think can exclude this one, or describe it in detail

File: site2/website/release-notes.md [hunk 4]
```diff
@@ -1,6 +1,109 @@
...
+* Completing connector configuration #4999
+* Add checkstyle validation and fix style violations in the common module #4989
+* Trim deleted entries after recover cursor #4987
+* Expose getLastMessageId method in ConsumerImpl #4911
+* Add a documentation page for metrics reference #4910
+* Merge Request for #4809: provide a convenient method for C++ client producer batch container #4885
+* Add schema admin api get schema info with schema version #4877
+* Return Message ID for send for cpp and cgo client #4811
+* Add the schema admin api #4800
+* Test cleanup and simplification #4799
```
Comment (codelipenghui) [inline site2/website/release-notes.md]: Shall we need to include the test enhancement in release node?

**Comment by wolfstudy** [comment]
@codelipenghui fixed done, PTAL again, thanks.

File: site2/website/release-notes.md [hunk 5]
```diff
@@ -1,6 +1,107 @@
...
+* Don't return DEFAULT_RACK if ZkBookieRackAffinityMapping can't resolve network location [#5067](https://github.com/apache/pulsar/pull/5067)
+* Reload zk cache asynchronously [#5049](https://github.com/apache/pulsar/pull/5049)
+* Add different cache flags to ConcurrentOpenLongPairRangeSet for size() and toString() [#5040](https://github.com/apache/pulsar/pull/5040)
+* Introduce number of threads in perf producer program [#5036](https://github.com/apache/pulsar/pull/5036)
+* Completing connector configuration [#4999](https://github.com/apache/pulsar/pull/4999)
+* Add checkstyle validation and fix style violations in the common module [#4989](https://github.com/apache/pulsar/pull/4989)
+* Trim deleted entries after recover cursor [#4987](https://github.com/apache/pulsar/pull/4987)
+* Expose getLastMessageId method in ConsumerImpl [#4911](https://github.com/apache/pulsar/pull/4911)
+* Add a documentation page for metrics reference [#4910](https://github.com/apache/pulsar/pull/4910)
+* Merge Request for #4809: provide a convenient method for C++ client producer batch container [#4885](https://github.com/apache/pulsar/pull/4885)
```
Comment (codelipenghui) [inline site2/website/release-notes.md]: Need `Merge Request for #4809` here?

File: site2/website/release-notes.md [hunk 6]
```diff
@@ -1,6 +1,107 @@
 
 ## Apache
 
+### 2.4.2 &mdash; 2019-12-04 <a id="2.4.2"></a>
+
+#### Fixes
+
+* Fixed don't set interrupt flag again after catching interrupt exception in Pulsar Client [#5643](https://github.com/apache/pulsar/pull/5643)
+* Fixed data is not deleted after expiration due to connected readers [#5621](https://github.com/apache/pulsar/pull/5621)
+* Fixed typo in BrokersBase file [#5596](https://github.com/apache/pulsar/pull/5596)
```
Comment (codelipenghui) [inline site2/website/release-notes.md]: No need to add the typo fix in to release note, right?

File: site2/website/release-notes.md [hunk 7]
```diff
@@ -1,6 +1,107 @@
 
 ## Apache
 
+### 2.4.2 &mdash; 2019-12-04 <a id="2.4.2"></a>
+
+#### Fixes
+
+* Fixed don't set interrupt flag again after catching interrupt exception in Pulsar Client [#5643](https://github.com/apache/pulsar/pull/5643)
```
Comment (codelipenghui) [inline site2/website/release-notes.md]: It's not a clear briefly description for this PR, i think you'd better to check the PR

File: site2/website/release-notes.md [hunk 8]
```diff
@@ -1,6 +1,107 @@
...
+* Fixed StringSchema static initialization  [#5445](https://github.com/apache/pulsar/pull/5445)
+* Fixed message corruption on OOM for batch messages [#5443](https://github.com/apache/pulsar/pull/5443)
+* Fixed couple functions related integration tests [#5434](https://github.com/apache/pulsar/pull/5434)
+* Fixed bug that namespace policies does not take effect due to NPE [#5408](https://github.com/apache/pulsar/pull/5408)
+* Fixed race condition : Failed to read-more entries on dispatcher [#5391](https://github.com/apache/pulsar/pull/5391)
+* Fixed potential deadlock that can occur in addConsumer [#5371](https://github.com/apache/pulsar/pull/5371)
+* Fixed proxy to be able to re-send request body [#5361](https://github.com/apache/pulsar/pull/5361)
+* Fixed pulsar can't load the customized SerDe [#5357](https://github.com/apache/pulsar/pull/5357)
+* Fixed instability in Pulsar Function window integration test [#5337](https://github.com/apache/pulsar/pull/5337)
+* Fixed bk write failure part 2 [#5322](https://github.com/apache/pulsar/pull/5322)
```
Comment (codelipenghui) [inline site2/website/release-notes.md]: We'd better have a briefly description for the failure

File: site2/website/release-notes.md [hunk 9]
```diff
@@ -1,6 +1,107 @@
...
+* Fixed bad_weak_ptr error when closing producer [#5315](https://github.com/apache/pulsar/pull/5315)
+* Fixed typo in prometheus metrics [#5299](https://github.com/apache/pulsar/pull/5299)
```
Comment (codelipenghui) [inline site2/website/release-notes.md]: Is also a typo

File: site2/website/release-notes.md [hunk 10]
```diff
@@ -1,6 +1,107 @@
...
+* Fixed invalidate cache on zk-cache timeout [#5298](https://github.com/apache/pulsar/pull/5298)
+* Don't attempt to append on read-only cursor ledger [#5297](https://github.com/apache/pulsar/pull/5297)
```
Comment (codelipenghui) [inline site2/website/release-notes.md]: I think this PR is a enhancement,  Avoid persistent Unack Ranges while cursor recovering

File: site2/website/release-notes.md [hunk 11]
```diff
@@ -1,6 +1,107 @@
...
+* Fixed memory leak caused by not being executed ClientConnection destructor [#5286](https://github.com/apache/pulsar/pull/5286)
+* Fixed producer blocked after send an over size message while batch enabled [#5282](https://github.com/apache/pulsar/pull/5282)
+* Fixed race condition while triggering message redelivery after an ack-timeout event [#5276](https://github.com/apache/pulsar/pull/5276)
+* Fixed behavior when getting a key from functions state that doesn't exist [#5272](https://github.com/apache/pulsar/pull/5272)
+* Fixed Cmake to build _pulsar for osx #5263
+* Fixed client backoff #5261
```
Comment (codelipenghui) [inline site2/website/release-notes.md]: Need  a clear briefly description here

**Comment by codelipenghui** [comment]
@Anonymitaet Please help review the 2.4.2 release note

File: site2/website/release-notes.md [hunk 12]
```diff
@@ -1,6 +1,107 @@
...
+* Fixed memory leak caused by deadline_timer holding object reference [#5246](https://github.com/apache/pulsar/pull/5246)
+* Fixed in Message Deduplication that may cause incorrect client/broker interaction [#5243](https://github.com/apache/pulsar/pull/5243)
+* Fixed bug that fails to search namespace bundle due to NPE [#5191](https://github.com/apache/pulsar/pull/5191)
+* Fixed bug that message delivery stops after resetting cursor for failover subscription [#5185](https://github.com/apache/pulsar/pull/5185)
+* Fixed exception type check order bug [#5174](https://github.com/apache/pulsar/pull/5174)
+* Fixed spark receiver to account for all the consumer config options [#5152](https://github.com/apache/pulsar/pull/5152)
+* Fixed broker fails to start with function worker enabled and broker client using TLS [#5151](https://github.com/apache/pulsar/pull/5151)
+* Fixed deadlock when resetting cursor [#5136](https://github.com/apache/pulsar/pull/5136)
+* Fixed windowed functions were broken when we changed java function instance to use classloaders [#5117](https://github.com/apache/pulsar/pull/5117)
+* Fixed storage size always 0 without subscription #5108
```
Comment (Anonymitaet) [inline site2/website/release-notes.md]: need add a link for #5108?

**Comment by codelipenghui** [comment]
retest this please

**Comment by wolfstudy** [comment]
run cpp tests
run java8 tests

**Comment by wolfstudy** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run integration tests

**Comment by wolfstudy** [comment]
run intergation tests

**Comment by wolfstudy** [comment]
run integration tests

**Pull request merged**